Moya/ReactiveCocoa/Moya+ReactiveCocoa.swift

```diff
@@ -2,68 +2,76 @@ import Foundation
+                    
+                    let cancellableToken = self?.request(token) { data, statusCode, response, error in
+                        if let error = error {
+                            if let statusCode = statusCode {
+                                sendError(requestSink, NSError(domain: MoyaErrorDomain, code: statusCode, userInfo: [NSUnderlyingErrorKey: error as NSError]))
+                            } else {
+                                sendError(requestSink, error as NSError)
+                            }
                         } else {
-                            subscriber.sendError(error as NSError)
+                            if let data = data {
+                                sendNext(requestSink, MoyaResponse(statusCode: statusCode!, data: data, response: response))
+                            }
+                            sendCompleted(requestSink)
                         }
-                    } else {
-                        if let data = data {
-                            subscriber.sendNext(MoyaResponse(statusCode: statusCode!, data: data, response: response))
+                    }
+                    
+                    requestDisposable.addDisposable {
+                        if let weakSelf = self {
+                            objc_sync_enter(weakSelf)
+                            // Clear the inflight request
+                            weakSelf.inflightRequests[endpoint] = nil
+                            objc_sync_exit(weakSelf)
+                            // Cancel the request
+                            cancellableToken?.cancel()
                         }
-                        subscriber.sendCompleted()
                     }
                 }
                 
-                return RACDisposable { () -> Void in
+                /// starts the inner signal producer and store the created signal.
+                producer.startWithSignal { [weak self] signal, innerDisposable in
                     if let weakSelf = self {
                         objc_sync_enter(weakSelf)
-                        weakSelf.inflightRequests[endpoint] = nil
-                        cancellableToken?.cancel()
+                        weakSelf.inflightRequests[endpoint] = signal
                         objc_sync_exit(weakSelf)
+                        /// connect all events of the signal to the observer of this signal producer
+                        signal.observe(outerSink)
+                        outerDisposable.addDisposable(innerDisposable)
                     }
                 }
-            }.publish().autoconnect()
-            
-            if let weakSelf = self {
-                objc_sync_enter(weakSelf)
-                weakSelf.inflightRequests[endpoint] = signal
-                objc_sync_exit(weakSelf)
             }
-            
-            return signal
         }
     }
+    
+    public func request(token: T) -> RACSignal {
```

Comment:
It shouldn't be if the developer has their types set properly. For example if the developer just has something like this `let signal = provider.request(target)` then there will be an error. They would need to specify the type or chain it directly with a subscribe like so `provider.request(target).subscribeNext { _ in  }`. It's not really breaking but it might cause errors for some users.